### PR TITLE
Lock-voting

### PR DESCRIPTION
**Pull request by gavofyork**
This finishes up the changes I've been presenting at meetups recently.

### CODE REVIEW

File: srml/democracy/src/lib.rs [hunk 1]
```diff
@@ -292,51 +335,69 @@ impl<T: Trait> Module<T> {
...
+		let total_issuance = <balances::Module<T>>::total_issuance();
+		let approved = info.threshold.approved(approve, against, capital, total_issuance);
+		let lock_period = Self::public_delay();
+
+		for (a, vote) in Self::voters_for(index).into_iter()
+			.map(|a| (a.clone(), Self::vote_of((index, a))))
+			// ^^^ defensive only: all items come from `voters`; for an item to be in `voters` there must be a vote registered; qed
+			.filter(|&(_, vote)| vote.is_aye() == approved)
+		{
+			let locked_until = now + lock_period * T::BlockNumber::sa((vote.multiplier()) as u64);
```
Comment (bkchr) [inline srml/democracy/src/lib.rs]: Shouldn't the multiplier only be used if the referendum was approved?

Comment (rphmeier) [reply srml/democracy/src/lib.rs]: (answering this in the documentation of the module or function would be ideal)

File: srml/democracy/src/lib.rs [hunk 2]
```diff
@@ -218,30 +258,33 @@ impl<T: Trait> Module<T> {
...
 	/// Get the voters for the current proposal.
-	pub fn tally(ref_index: ReferendumIndex) -> (T::Balance, T::Balance) {
+	pub fn tally(ref_index: ReferendumIndex) -> (T::Balance, T::Balance, T::Balance) {
 		Self::voters_for(ref_index).iter()
-			.map(|a| (<balances::Module<T>>::total_balance(a), Self::vote_of((ref_index, a.clone())).unwrap_or(false)/*defensive only: all items come from `voters`; for an item to be in `voters` there must be a vote registered; qed*/))
-			.map(|(bal, vote)| if vote { (bal, Zero::zero()) } else { (Zero::zero(), bal) })
-			.fold((Zero::zero(), Zero::zero()), |(a, b), (c, d)| (a + c, b + d))
+			.map(|voter| (
+				<balances::Module<T>>::total_balance(voter),
+				Self::vote_of((ref_index, voter.clone())),
```
Comment (andresilva) [inline srml/democracy/src/lib.rs]: I know this get shouldn't fail but consider documenting that the default value of `Vote` is a nay vote.

File: srml/democracy/src/lib.rs (hunk 1 above)
Comment (andresilva) [reply srml/democracy/src/lib.rs]: If it was rejected only those who voted against get their bond locked. Maybe this merits a comment in the `filter` above.

File: srml/democracy/src/lib.rs [hunk 3]
```diff
@@ -186,8 +225,9 @@ decl_storage! {
 		/// Get the voters for the current proposal.
 		pub VotersFor get(voters_for): map ReferendumIndex => Vec<T::AccountId>;
 
-		/// Get the vote, if Some, of `who`.
-		pub VoteOf get(vote_of): map (ReferendumIndex, T::AccountId) => Option<bool>;
+		/// Get the vote, if Some, of `who`. The is the additional periods that the voter offer to lock their
+		/// vote tokens in the case of success.
```
Comment (andresilva) [inline srml/democracy/src/lib.rs]: Document that it returns a default value if the `who` didn't vote.

File: srml/democracy/src/vote_threshold.rs [hunk 4]
```diff
@@ -35,7 +35,7 @@ pub trait Approved<Balance> {
 	/// Given `approve` votes for and `against` votes against from a total electorate size of
 	/// `electorate` (`electorate - (approve + against)` are abstainers), then returns true if the
 	/// overall outcome is in favour of approval.
-	fn approved(&self, approve: Balance, against: Balance, electorate: Balance) -> bool;
+	fn approved(&self, approve: Balance, against: Balance, voters: Balance, electorate: Balance) -> bool;
```
Comment (andresilva) [inline srml/democracy/src/vote_threshold.rs]: Docs should be updated (also below in the `impl`).

File: srml/democracy/src/lib.rs (hunk 1 above)
Comment (gavofyork) [reply srml/democracy/src/lib.rs]: The logic is correct; I added a few comments and a link to my presentation.

File: srml/democracy/src/vote_threshold.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (gavofyork) [reply srml/democracy/src/vote_threshold.rs]: The balance of voting tokens.

File: srml/democracy/src/lib.rs [hunk 5]
```diff
@@ -49,6 +49,35 @@ pub use vote_threshold::{Approved, VoteThreshold};
...
+
+/// A number of lock periods, plus a vote, one way or the other.
+#[derive(Encode, Decode, Copy, Clone, Eq, PartialEq, Default)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Vote(i8);
+
+impl Vote {
+	/// Create a new instance.
+	pub fn new(aye: bool, multiplier: LockPeriods) -> Self {
+		let m = multiplier.max(1) - 1;
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply srml/democracy/src/lib.rs]: Don't think it can.

`-128.max(1) - 1 == 0`

**Pull request merged**